src/object/notification.c

```diff
@@ -259,6 +259,9 @@ void completeSignal(notification_t *ntfnPtr, tcb_t *tcb)
         badge = notification_ptr_get_ntfnMsgIdentifier(ntfnPtr);
         setRegister(tcb, badgeRegister, badge);
         notification_ptr_set_state(ntfnPtr, NtfnState_Idle);
+#ifdef CONFIG_KERNEL_MCS
+        maybeDonateSchedContext(tcb, ntfnPtr);
```

Comment:
nice

Reply:
this commit can go in

Reply:
I left this commit out of https://github.com/seL4/seL4/pull/218 at the moment, because we still aren't sure about the next commit that returns the SC if the thread blocks.  My understanding of the current behavior:
- A notification will only donate its SC if a passive thread waits on it, if the notification is bound and the passive server waits on an endpoint, the notification will be delivered but the SC won't be donated.  This will cause the passive thread to be made runnable but not schedulable and it will no longer be in any queue.
- A donated notification SC will only be returned if the thread waits on the notification again.

My preference is that we change the donation behavior of bound notifications in the same PR as we change how they get returned.

Reply:
The initial design was the opposite of the above point #1 - we want to be able to have passive servers which receive notifications.

so the semantics should be as follows:

* passive threads receives SCs any notification  notification objects if they receive a notification, regardless of whether this is via a bound endpoint or directly blocking on the notification
* If a passive thread _T_ is running on _SC_ which is received from notification _n_ then it should be returned to _n_ if _T_ blocks on its bound endpoint, or if _T_ blocks on _n_.

Reply:
But endpoints aren't bound to threads, notification s are. Are you suggesting that we add bound endpoints to threads?